citizennet/purescript-ocelot
# JS Interfaces

## What does this pull request do?

This PR introduces support for JavaScript interfaces to Halogen components. In short, these allow teams that do not use PureScript to still leverage PureScript components, without needing to understand the PureScript implementation and by writing idiomatic JavaScript. This means that it's possible for one team to transition to PureScript while continuing to share code with other teams.

Notably, this PR allows for idiomatic JS to consume a stream of outputs from the component as well as trigger requests and actions in the component. It is possible to dispose of the component by destroying the element it is mounted within.

## Where should the reviewer start?

Review the new typeahead interface to see how we transform the PureScript representation into one more accessible to JavaScript code. Then, review the `test/Interop` directory for a working example of the Ocelot typeahead mounted from JS.

## How should this be manually tested?

Verify that you can build and control the component in the `test/Interop` directory.

## Other Notes:

This should be considered a first-cut implementation suitable for mounting a single component, but perhaps not the final approach we'll take to these interfaces. This requires that any consumer have a module bundler working, like Webpack.
